### PR TITLE
us: add list users API to v5 [v5supportListUsers]

### PR DESCRIPTION
**Pull request by DvirCyberArk**
Same as V4 we would like to align API, expose list users resource

### CODE REVIEW

File: conjur-api/Resource.cs [hunk 1]
```diff
@@ -110,5 +110,7 @@ public bool Check(string privilege)
                 offset += (uint)resultList.Count;
             } while (resultList.Count > 0);
         }
+
+        protected static string IdToName (string id, string account, ResourceKind kind) => id.Substring (id.IndexOf ($"{account}:{kind}:") + 1);
```
Comment (dividedmind) [inline conjur-api/Resource.cs]: Is there any reason not to just pass the id and return the last part? 
Also, this function would be clearer using full function definition syntax.

File: conjur-api/Client.Methods.cs [hunk 2]
```diff
@@ -36,6 +36,28 @@ public IEnumerable<Variable> ListVariables(string query = null)
             return Conjur.Variable.List(this, query);
         }
 
+        /// <summary>
+        /// Create an onject representing the User with this specified name.
+        /// </summary>
+        /// <returns>The user.</returns>
+        /// <param name="name">Name.</param>
+        /// <returns>User object.</returns>
+        /// <seealso cref="User()"/>
```
Comment (dividedmind) [inline conjur-api/Client.Methods.cs]: Please pay closer attention to docstrings. This one has two `<returns>`. Also they should be more verbose than just repeating the method signature.

File: conjur-api/Client.Methods.cs [hunk 3]
```diff
@@ -36,6 +36,28 @@ public IEnumerable<Variable> ListVariables(string query = null)
...
+        public User User(string name)
+        {
+            return new User(this, name);
+        }
+
+        /// <summary>
+        /// Lists the users.
+        /// </summary>
+        /// <returns>The variabßles.</returns>
+        /// <param name="query">Query.</param>
```
Comment (dividedmind) [inline conjur-api/Client.Methods.cs]: This docstring could use improvement.

File: conjur-api/User.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    /// Conjur user reference
+    /// </summary>
+    /// a user represents resource for a human identity
+    public class User : Resource
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:Conjur.User"/> class.
+        /// </summary>
+        /// <param name="client">Client.</param>
+        /// <param name="name">Name.</param>
```
Comment (dividedmind) [inline conjur-api/User.cs]: Please make these more verbose. (OTOH these are internal, so arguably docstrings aren't that important here.)

File: conjur-api/Resource.cs (hunk 1 above)
Comment (dividedmind) [inline conjur-api/Resource.cs]: Is there any reason not to just pass the id and return the last part? 
Also, this function would be clearer using full function definition syntax.

Comment (acher82) [reply conjur-api/Resource.cs]: Until now my impression is that colon is a valid symbol of resource name.
The next policy will be uploaded successfully
```
- !user
  id: abc:def
```
So simple splitting by colon and returning the last part will be definitely a bug.

Comment (dividedmind) [reply conjur-api/Resource.cs]: Well, you still don't need to pass account and kind explicitly. Here's how conjur itself does it: `self.id.split(':', 3)[2]`, cf.
https://github.com/cyberark/conjur/blob/37ddd2d51b6612edc52b430d0fd20174b501e08e/app/models/has_id.rb#L11

I believe C# equivalent would be almost the same, ie. `id.Split(":", 3)[2]`.

Comment (acher82) [reply conjur-api/Resource.cs]: Here I disagree with you.
Split does not make any sense here. Current implementation is self explained - resource id combined from account, kind and name. I think that it's important.
By the way colon is also valid symbol for kind (and may be for account also). I checked that the next policy works well for me:
```
- !resource &north_gate
  kind: ga:te
  id: north
```

File: conjur-api/Variable.cs [hunk 5]
```diff
@@ -60,7 +60,7 @@ public void AddSecret(string val)
         internal static IEnumerable<Variable> List(Client client, string query = null)
         {
             Func<ResourceMetadataֿ, Variable> newInst = (searchRes) => new Variable(client, IdToName(searchRes.Id, client.GetAccountName(), ResourceKind.variable));
-            return ListResources(client, ResourceKind.variable, newInst, query);
+            return ListResources<Variable, ResourceMetadataֿ >(client, ResourceKind.variable, newInst, query);
```
[the remark this answers is not in the document]
Comment (DvirCyberArk) [reply conjur-api/Variable.cs]: It compiles and even pass build.sh :)
Don't understand those symbols also at "ResourceMetadata," a row above - when you try to copy it to textbox here those symbols disappear

File: conjur-api/User.cs [hunk 6]
```diff
@@ -1,36 +1,40 @@
-﻿namespace Conjur
+﻿// <copyright file="User.cs" company="Conjur Inc.">
+//     Copyright (c) 2016 Conjur Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (DvirCyberArk) [reply conjur-api/User.cs]: You are correct but I didn't want to change those until legal or other authority will align (all?) classes copyrights in a separate pull request

File: conjur-api/Variable.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (DvirCyberArk) [reply conjur-api/Variable.cs]: Noticed your catch at previous pull request (listVariables) changed the class name.

File: conjur-api/Resource.cs (hunk 1 above)
Comment (acher82) [reply conjur-api/Resource.cs]:
```
- !resource
  kind: ab:cd
  id: ef:gh
```
`conjur policy load root resource.yml`
Policy loaded successfully.
`curl -H "$(conjur authn authenticate -H)" http://conjur/resources/quick-start/ab%3Acd/ef%3Agh | jq .`
and
`curl -H "$(conjur authn authenticate -H)" http://conjur/resources/quick-start/ab/cd%3Aef%3Agh | jq .`
both return the right output
```
{
  "created_at": "2018-06-28T19:01:18.401+00:00",
  "id": "quick-start:ab:cd:ef:gh",
  "owner": "quick-start:user:admin",
  "policy": "quick-start:policy:root",
  "permissions": [],
  "annotations": []
}
```
So I still think that we need to use non-split parsing implementation.

**Comment by DvirCyberArk** [comment]
Hi,
Please review this merged change (merged V5support to this branch), also in the pull request I have changed the base branch to be V5support instead of v5supportListVariables

@acher82 @jvanderhoof

**Pull request merged**